File: src/agent/riflettIntentClassifier.ts
import type { MemoryRecord } from '@/agent/memory';
import { toTitleCase } from '@/utils/strings';
import { ContextWindow } from './contextWindow';

export type RiflettIntentLabel =
  | 'conversational'
  | 'entry_create'
  | 'entry_discuss'
  | 'entry_append'
  | 'command'
  | 'search_query';

export interface ClassificationCandidate {
  label: RiflettIntentLabel;
  confidence: number;
}

export interface ClassificationMeta {
  label: RiflettIntentLabel;
  confidence: number;
  reasons: string[];
  targetEntryId?: string | null;
  targetEntryType?: string | null;
  duplicateMatch?: {
    id: string;
    score: number;
    text: string;
    kind: string;
  } | null;
  topCandidates: ClassificationCandidate[];
}

const COMMAND_PATTERN = /^\s*\//;
const SEARCH_PHRASES = [
  'find',
  'show me',
  'search for',
  'when did i',
  'where did i',
  'what did i write',
  'look up',
  'list',
];
const ADDITIVE_PHRASES = [
  'also',
  'update',
  'another',
  'forgot',
  'in addition',
  'plus',
  'adding',
  'one more thing',
];
const SAVE_PHRASES = [
  'save',
  'log',
  'capture',
  'remember',
  'write this down',
  'note that',
  'journal',
  'record',
];
const TEMPORAL_MARKERS = [
  'today',
  'tonight',
  'this morning',
  'this afternoon',
  'this evening',
  'yesterday',
  'earlier',
  'just now',
  'right now',
  'tomorrow',
  'monday',
  'tuesday',
  'wednesday',
  'thursday',
  'friday',
  'saturday',
  'sunday',
];
const ACTION_VERBS = [
  'finished',
  'completed',
  'did',
  'ran',
  'talked',
  'spoke',
  'met',
  'decided',
  'started',
  'launched',
  'sent',
  'emailed',
  'called',
  'worked',
  'wrote',
  'built',
  'shipped',
];
const PRONOUN_ANCHORS = /\b(it|this|that|the goal|the entry|that goal|this entry|that plan|the plan|that project|this project)\b/i;
const QUESTION_MARK = /\?/;
const TIME_REGEX = /\b(\d{1,2}(:\d{2})?\s?(am|pm))\b/i;

const containsPhrase = (text: string, phrases: string[]): boolean =>
  phrases.some((phrase) => text.includes(phrase));

const hasTemporalMarker = (text: string): boolean =>
  containsPhrase(text, TEMPORAL_MARKERS) || TIME_REGEX.test(text);

const hasSaveLanguage = (text: string): boolean => containsPhrase(text, SAVE_PHRASES);

const hasAdditiveLanguage = (text: string): boolean => containsPhrase(text, ADDITIVE_PHRASES);

const hasActionLanguage = (text: string): boolean =>
  ACTION_VERBS.some((verb) => text.includes(verb));

const isSearchQuery = (text: string): boolean => containsPhrase(text, SEARCH_PHRASES);

const baseCandidate = (label: RiflettIntentLabel, confidence: number): ClassificationCandidate => ({
  label,
  confidence,
});

const DUPLICATE_THRESHOLD = 0.85;

function pickDuplicate(records: MemoryRecord[]): MemoryRecord | null {
  return records.find((record) => record.score >= DUPLICATE_THRESHOLD) ?? null;
}

function mapKindToEntryType(kind?: string | null): string | null {
  if (!kind) return null;
  const normalized = kind.toLowerCase();
  if (normalized.includes('goal')) return 'goal';
  if (normalized.includes('event')) return 'schedule';
  if (normalized.includes('entry') || normalized.includes('journal')) return 'journal';
  return null;
}

export function classifyRiflettIntent(params: {
  text: string;
  contextRecords: MemoryRecord[];
}): ClassificationMeta {
  const trimmed = params.text.trim();
  const lower = trimmed.toLowerCase();
  const reasons: string[] = [];
  const contextSnapshot = ContextWindow.snapshot();
  const recentMessages = ContextWindow.recent();

  if (!trimmed) {
    return {
      label: 'conversational',
      confidence: 0.6,
      reasons: ['Empty message defaults to conversational'],
      topCandidates: [baseCandidate('conversational', 0.6)],
    };
  }

  if (COMMAND_PATTERN.test(trimmed)) {
    reasons.push('Slash-prefixed command detected');
    return {
      label: 'command',
      confidence: 0.99,
      reasons,
      topCandidates: [baseCandidate('command', 0.99)],
    };
  }

  if (isSearchQuery(lower)) {
    reasons.push('Search verb detected');
    const confidence = lower.length > 24 ? 0.94 : 0.9;
    return {
      label: 'search_query',
      confidence,
      reasons,
      topCandidates: [
        baseCandidate('search_query', confidence),
        baseCandidate('conversational', 0.7),
      ],
    };
  }

  const duplicate = pickDuplicate(params.contextRecords);
  const duplicateMeta = duplicate
    ? {
        id: duplicate.id,
        score: duplicate.score,
        text: duplicate.text,
        kind: duplicate.kind,
      }
    : null;

  const hasAdditive = hasAdditiveLanguage(lower);
  const inWindow = Boolean(contextSnapshot?.isActive);
  const hasPronounAnchor = PRONOUN_ANCHORS.test(trimmed);

  if (duplicateMeta && hasAdditive) {
    reasons.push('High-similarity memory match with additive language');
    const entryType = mapKindToEntryType(duplicateMeta.kind);
    return {
      label: 'entry_append',
      confidence: 0.91,
      reasons,
      targetEntryId: duplicateMeta.id,
      targetEntryType: entryType,
      duplicateMatch: duplicateMeta,
      topCandidates: [
        baseCandidate('entry_append', 0.91),
        baseCandidate('entry_discuss', 0.8),
        baseCandidate('entry_create', 0.6),
      ],
    };
  }

  if (inWindow && hasPronounAnchor) {
    reasons.push('Within entry context window with pronoun reference');
    return {
      label: 'entry_discuss',
      confidence: 0.96,
      reasons,
      targetEntryId: contextSnapshot?.entryId ?? null,
      targetEntryType: contextSnapshot?.entryType ?? null,
      duplicateMatch: duplicateMeta,
      topCandidates: [
        baseCandidate('entry_discuss', 0.96),
        baseCandidate('entry_append', hasAdditive ? 0.82 : 0.7),
        baseCandidate('conversational', 0.68),
      ],
    };
  }

  if (duplicateMeta && !hasAdditive) {
    reasons.push('High-similarity memory match without clear additive cue');
  }

  const hasSave = hasSaveLanguage(lower);
  const hasTemporal = hasTemporalMarker(lower);
  const hasAction = hasActionLanguage(lower);
  const isQuestion = QUESTION_MARK.test(trimmed);
  const wordCount = trimmed.split(/\s+/).filter(Boolean).length;

  if (!isQuestion && (hasSave || (hasTemporal && hasAction) || wordCount > 25)) {
    const createConfidence = Math.min(
      0.86 + (hasSave ? 0.08 : 0) + (hasTemporal ? 0.03 : 0) + (wordCount > 60 ? 0.03 : 0),
      0.97,
    );
    reasons.push('Declarative content suitable for structured capture');
    if (hasSave) reasons.push('Explicit save/log intent detected');
    if (hasTemporal) reasons.push('Temporal marker detected');
    if (hasAction) reasons.push('Concrete action verb detected');

    return {
      label: 'entry_create',
      confidence: createConfidence,
      reasons,
      targetEntryId: null,
      duplicateMatch: duplicateMeta,
      topCandidates: [
        baseCandidate('entry_create', createConfidence),
        baseCandidate('entry_append', duplicateMeta ? 0.78 : 0.62),
        baseCandidate('conversational', 0.6),
      ],
    };
  }

  if (hasAdditive && duplicateMeta) {
    reasons.push('Additive phrasing referencing prior context');
    return {
      label: 'entry_append',
      confidence: 0.88,
      reasons,
      targetEntryId: duplicateMeta.id,
      targetEntryType: mapKindToEntryType(duplicateMeta.kind),
      duplicateMatch: duplicateMeta,
      topCandidates: [
        baseCandidate('entry_append', 0.88),
        baseCandidate('entry_discuss', 0.72),
        baseCandidate('conversational', 0.65),
      ],
    };
  }

  if (inWindow && !isQuestion) {
    reasons.push('Context window active; defaulting follow-up to discuss');
    return {
      label: 'entry_discuss',
      confidence: 0.78,
      reasons,
      targetEntryId: contextSnapshot?.entryId ?? null,
      targetEntryType: contextSnapshot?.entryType ?? null,
      duplicateMatch: duplicateMeta,
      topCandidates: [
        baseCandidate('entry_discuss', 0.78),
        baseCandidate('entry_create', 0.6),
        baseCandidate('conversational', 0.58),
      ],
    };
  }

  if (duplicateMeta && recentMessages.some((m) => m.text.toLowerCase().includes(duplicateMeta.text.slice(0, 20).toLowerCase()))) {
    reasons.push('Recent message references similar content; favour append');
    return {
      label: 'entry_append',
      confidence: 0.8,
      reasons,
      targetEntryId: duplicateMeta.id,
      targetEntryType: mapKindToEntryType(duplicateMeta.kind),
      duplicateMatch: duplicateMeta,
      topCandidates: [
        baseCandidate('entry_append', 0.8),
        baseCandidate('entry_discuss', 0.7),
        baseCandidate('conversational', 0.65),
      ],
    };
  }

  const conversationalConfidence = isQuestion ? 0.9 : 0.82;
  if (isQuestion) reasons.push('Question format leaning conversational');
  else reasons.push('Defaulting to reflective conversational mode');

  return {
    label: 'conversational',
    confidence: conversationalConfidence,
    reasons,
    targetEntryId: contextSnapshot?.isActive ? contextSnapshot.entryId : null,
    targetEntryType: contextSnapshot?.entryType ?? null,
    duplicateMatch: duplicateMeta,
    topCandidates: [
      baseCandidate('conversational', conversationalConfidence),
      baseCandidate('search_query', isQuestion ? 0.58 : 0.4),
      baseCandidate('entry_create', 0.35),
    ],
  };
}

export function toNativeLabel(label: RiflettIntentLabel): string {
  return toTitleCase(label.replace(/_/g, ' '));
}

File: services/python-ai/app/agent/riflett_intent_classifier.py
import re
from typing import Dict, List, Optional

from .context_window import recent as recent_messages, snapshot as window_snapshot
from .memory import MemoryRecord
from .utils.strings import to_title_case


RiflettIntentLabel = [
    "conversational",
    "entry_create",
    "entry_discuss",
    "entry_append",
    "command",
    "search_query",
]


ClassificationCandidate = Dict[str, float]

COMMAND_PATTERN = re.compile(r"^\s*/")
SEARCH_PHRASES = [
    "find",
    "show me",
    "search for",
    "when did i",
    "where did i",
    "what did i write",
    "look up",
    "list",
]
ADDITIVE_PHRASES = [
    "also",
    "update",
    "another",
    "forgot",
    "in addition",
    "plus",
    "adding",
    "one more thing",
]
SAVE_PHRASES = [
    "save",
    "log",
    "capture",
    "remember",
    "write this down",
    "note that",
    "journal",
    "record",
]
TEMPORAL_MARKERS = [
    "today",
    "tonight",
    "this morning",
    "this afternoon",
    "this evening",
    "yesterday",
    "earlier",
    "just now",
    "right now",
    "tomorrow",
    "monday",
    "tuesday",
    "wednesday",
    "thursday",
    "friday",
    "saturday",
    "sunday",
]
ACTION_VERBS = [
    "finished",
    "completed",
    "did",
    "ran",
    "talked",
    "spoke",
    "met",
    "decided",
    "started",
    "launched",
    "sent",
    "emailed",
    "called",
    "worked",
    "wrote",
    "built",
    "shipped",
]
PRONOUN_ANCHORS = re.compile(r"\b(it|this|that|the goal|the entry|that goal|this entry|that plan|the plan|that project|this project)\b", re.I)
QUESTION_MARK = re.compile(r"\?")
TIME_REGEX = re.compile(r"\b(\d{1,2}(:\d{2})?\s?(am|pm))\b", re.I)

DUPLICATE_THRESHOLD = 0.85
DUPLICATE_MIN_PREFIX_LENGTH = 20


def _contains_phrase(text: str, phrases: List[str]) -> bool:
    return any(phrase in text for phrase in phrases)


def _has_temporal_marker(text: str) -> bool:
    return _contains_phrase(text, TEMPORAL_MARKERS) or bool(TIME_REGEX.search(text))


def _has_save_language(text: str) -> bool:
    return _contains_phrase(text, SAVE_PHRASES)


def _has_additive_language(text: str) -> bool:
    return _contains_phrase(text, ADDITIVE_PHRASES)


def _has_action_language(text: str) -> bool:
    return any(verb in text for verb in ACTION_VERBS)


def _is_search_query(text: str) -> bool:
    return _contains_phrase(text, SEARCH_PHRASES)


def _base_candidate(label: str, confidence: float) -> Dict[str, float]:
    return {"label": label, "confidence": confidence}


def _pick_duplicate(records: List[MemoryRecord]) -> Optional[MemoryRecord]:
    return next((record for record in records if record.score >= DUPLICATE_THRESHOLD), None)


def _map_kind_to_entry_type(kind: Optional[str]) -> Optional[str]:
    if not kind:
        return None
    normalized = kind.lower()
    if "goal" in normalized:
        return "goal"
    if "event" in normalized or "schedule" in normalized:
        return "schedule"
    if "entry" in normalized or "journal" in normalized:
        return "journal"
    return None


def classify_riflett_intent(text: str, context_records: List[MemoryRecord]):
    trimmed = text.strip()
    lower = trimmed.lower()
    reasons: List[str] = []
    context_snapshot = window_snapshot()
    recent = recent_messages()

    if not trimmed:
        return {
            "label": "conversational",
            "confidence": 0.6,
            "reasons": ["Empty message defaults to conversational"],
            "topCandidates": [_base_candidate("conversational", 0.6)],
        }

    if COMMAND_PATTERN.search(trimmed):
        reasons.append("Slash-prefixed command detected")
        return {
            "label": "command",
            "confidence": 0.99,
            "reasons": reasons,
            "topCandidates": [_base_candidate("command", 0.99)],
        }

    if _is_search_query(lower):
        reasons.append("Search verb detected")
        confidence = 0.94 if len(lower) > 24 else 0.9
        return {
            "label": "search_query",
            "confidence": confidence,
            "reasons": reasons,
            "topCandidates": [
                _base_candidate("search_query", confidence),
                _base_candidate("conversational", 0.7),
            ],
        }

    duplicate = _pick_duplicate(context_records)
    duplicate_meta = (
        {"id": duplicate.id, "score": duplicate.score, "text": duplicate.text, "kind": duplicate.kind}
        if duplicate
        else None
    )

    has_additive = _has_additive_language(lower)
    in_window = bool(context_snapshot and context_snapshot.get("isActive"))
    has_pronoun_anchor = bool(PRONOUN_ANCHORS.search(trimmed))

    if duplicate_meta and has_additive:
        reasons.append("High-similarity memory match with additive language")
        entry_type = _map_kind_to_entry_type(duplicate_meta["kind"])
        return {
            "label": "entry_append",
            "confidence": 0.91,
            "reasons": reasons,
            "targetEntryId": duplicate_meta["id"],
            "targetEntryType": entry_type,
            "duplicateMatch": duplicate_meta,
            "topCandidates": [
                _base_candidate("entry_append", 0.91),
                _base_candidate("entry_discuss", 0.8),
                _base_candidate("entry_create", 0.6),
            ],
        }

    if in_window and has_pronoun_anchor:
        reasons.append("Within entry context window with pronoun reference")
        return {
            "label": "entry_discuss",
            "confidence": 0.96,
            "reasons": reasons,
            "targetEntryId": context_snapshot.get("entryId") if context_snapshot else None,
            "targetEntryType": context_snapshot.get("entryType") if context_snapshot else None,
            "duplicateMatch": duplicate_meta,
            "topCandidates": [
                _base_candidate("entry_discuss", 0.96),
                _base_candidate("entry_append", 0.82 if has_additive else 0.7),
                _base_candidate("entry_create", 0.6),
            ],
        }

    if duplicate_meta and not has_additive:
        reasons.append("High-similarity memory match without clear additive cue")

    has_save = _has_save_language(lower)
    has_temporal = _has_temporal_marker(lower)
    has_action = _has_action_language(lower)
    is_question = bool(QUESTION_MARK.search(trimmed))
    word_count = len([word for word in re.split(r"\s+", trimmed) if word])

    if not is_question and (has_save or (has_temporal and has_action) or word_count > 25):
        create_confidence = min(
            0.86 + (0.08 if has_save else 0) + (0.03 if has_temporal else 0) + (0.03 if word_count > 60 else 0),
            0.97,
        )
        reasons.append("Declarative content suitable for structured capture")
        if has_save:
            reasons.append("Explicit save/log intent detected")
        if has_temporal:
            reasons.append("Temporal marker detected")
        if has_action:
            reasons.append("Concrete action verb detected")

        return {
            "label": "entry_create",
            "confidence": create_confidence,
            "reasons": reasons,
            "targetEntryId": None,
            "duplicateMatch": duplicate_meta,
            "topCandidates": [
                _base_candidate("entry_create", create_confidence),
                _base_candidate("entry_append", 0.78 if duplicate_meta else 0.62),
                _base_candidate("conversational", 0.6),
            ],
        }

    if has_additive and duplicate_meta:
        reasons.append("Additive phrasing referencing prior context")
        return {
            "label": "entry_append",
            "confidence": 0.88,
            "reasons": reasons,
            "targetEntryId": duplicate_meta["id"],
            "targetEntryType": _map_kind_to_entry_type(duplicate_meta["kind"]),
            "duplicateMatch": duplicate_meta,
            "topCandidates": [
                _base_candidate("entry_append", 0.88),
                _base_candidate("entry_discuss", 0.72),
                _base_candidate("conversational", 0.65),
            ],
        }

    if in_window and not is_question:
        reasons.append("Context window active; defaulting follow-up to discuss")
        return {
            "label": "entry_discuss",
            "confidence": 0.78,
            "reasons": reasons,
            "targetEntryId": context_snapshot.get("entryId") if context_snapshot else None,
            "targetEntryType": context_snapshot.get("entryType") if context_snapshot else None,
            "duplicateMatch": duplicate_meta,
            "topCandidates": [
                _base_candidate("entry_discuss", 0.78),
                _base_candidate("entry_create", 0.6),
                _base_candidate("conversational", 0.58),
            ],
        }

    if duplicate_meta and len(duplicate_meta["text"]) >= DUPLICATE_MIN_PREFIX_LENGTH:
        duplicate_prefix = duplicate_meta["text"][:DUPLICATE_MIN_PREFIX_LENGTH].lower()
        recent_match = any(duplicate_prefix in message.text.lower() for message in recent)
        if recent_match:
            reasons.append("Recent message references similar content; favour append")
            return {
                "label": "entry_append",
                "confidence": 0.8,
                "reasons": reasons,
                "targetEntryId": duplicate_meta["id"],
                "targetEntryType": _map_kind_to_entry_type(duplicate_meta["kind"]),
                "duplicateMatch": duplicate_meta,
                "topCandidates": [
                    _base_candidate("entry_append", 0.8),
                    _base_candidate("entry_discuss", 0.7),
                    _base_candidate("conversational", 0.65),
                ],
            }

    conversational_confidence = 0.9 if is_question else 0.82
    if is_question:
        reasons.append("Question format leaning conversational")
    else:
        reasons.append("Defaulting to reflective conversational mode")

    return {
        "label": "conversational",
        "confidence": conversational_confidence,
        "reasons": reasons,
        "targetEntryId": context_snapshot.get("entryId") if context_snapshot and context_snapshot.get("isActive") else None,
        "targetEntryType": context_snapshot.get("entryType") if context_snapshot else None,
        "duplicateMatch": duplicate_meta,
        "topCandidates": [
            _base_candidate("conversational", conversational_confidence),
            _base_candidate("search_query", 0.58 if is_question else 0.4),
            _base_candidate("entry_create", 0.35),
        ],
    }


def to_native_label(label: str) -> str:
    return to_title_case(label.replace("_", " "))

